Install public headers during build

maleadt opened the pull request:
Instead of relying on `make install` to gather public headers and put them in the installation directory, this PR makes regular build in `src` already gather the headers in `$(BUILDDIR)/usr/include/julia`, with `make install` just copying that directory.

Fixes #18993, making it possible to build against julia (using flags from `julia-config.jl`) without having to run `make install`.

maleadt committed:
Install public headers during regular compilation already.

This makes it possible to build against (ie. include julia.h) a non-installed build tree.

tkelman commented on src/Makefile:
```diff
@@ -63,7 +63,18 @@ SRCS += anticodegen
+PUBLIC_HEADERS += $(build_includedir)/tree.h
+endif
+HEADERS += $(LIBUV_INC)/uv.h
+PUBLIC_HEADERS += $(LIBUV_INC)/uv.h
```
don't think this one file is always self-contained, I think the `*` is necessary

maleadt replied:
OK, updated.

tkelman commented on src/Makefile:
```diff
@@ -123,6 +134,14 @@ $(BUILDDIR)/%.o: $(SRCDIR)/%.cpp $(SRCDIR)/llvm-version.h $(HEADERS) $(LLVM_CONF
+define install_public_header
+TARGET = $(build_includedir)/julia/$(notdir $(1))
+$$(TARGET): $(1)
+	$(INSTALL_F) $$^ $(build_includedir)/julia/
```
something will probably need to create this as a directory first, otherwise won't the first header that gets installed end up named `$(build_includedir)/julia` as a file?

maleadt replied:
It relies on the toplevel Makefile calling `dir_target` on `DIRS`, but it's not registered as a dependency. I'll add an explicit one.

maleadt committed:
Add a dependency on the public header directory.

maleadt committed:
Also install julia-config.jl as part of regular build.

This puts julia-config.jl in a deterministic place relative to JULIA_HOME,
regardless of the type of build (binary dist, in-tree or out-of-tree source build).

maleadt committed:
Make the public headers a dependency of the libjulia targets.

When doing 'make install', we're building the 'julia-release' target and not the 'release' one.

tkelman commented:
xref https://github.com/JuliaLang/julia/pull/17745 we can actually do without some of the ones from src/support
